key-derivation.md

```diff
@@ -72,13 +68,26 @@ credentialMac = SHA256HMAC(seedKey XOR credentialMacMask, rpId || version || uni
 
 Derive the private key **`es256SPrivateKey`**  from the `seedKey` on the authenticator, the `rpId` from the relying party, and the `credentialMac` field from the `credentialId` (which, as a MAC, effectively encapsulates the three other fields of `credentialId`.
 
+
+Following [NIST FIPS 186.4](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-4.pdf) Section Section B.4.2,
+derive the secret key by testing randomly-generated 32-byte candidate sequences c_0, c_1, ... c_i of the form 
+
 ```
-es256SPrivateKey = SHA256HMAC(seedKey XOR es256SPrivateKeyMask, rpId || credentialMac)
+c_i = LE SHA256HMAC(seedKey XOR (es256SPrivateKeyMask ^ i), rpId || credentialMac)
 ```
 
-To derive the public key to send to return, use the existing deterministic algorithm for calculating ES public keys from private keys.
+Where c_i and i are treated as converted between byte arrays and numeric values using little endian representation.
 
+So, the algorithm is:
 
+```
+i = 0;
+c = p;
+for (i=0; c > p-2; i++) {
+  c = LE SHA256HMAC(seedKey XOR (es256SPrivateKeyMask ^ i), rpId || credentialMac)
+}
+d = C + 1  // public key Q = dG
+```
```

Comment:
This is very complicated: I can't keep it in my head easily; and I see no advantage over just hashing. Motivation?

Reply:
It was you who asked asked that if there were a standard, that we use it.  For dealing with the case where a candidate key is 0, comparing to 0 and < p (your method) feels more elegant than comparing <= p-2 and then incrementing (their method)  because their method requires keeping p-1 or p-2 in memory.  It doesn't seem sufficiently less elegant to justify avoiding the standard.

The notation is from the standard and the algorithm is just implementation of the one in the standard, which seems harder to follow than the 1 line for loop above.   The sample implementation of the comparisons and loop structure doesn't really matter since the standard is definitive. Since it's just _an_ implementation of a standard, the only thing we have to agree on is the bit stream.  (The i=0 line is actually superfluous and should be removed.  I introduced c=p to ensure the first comparison always fails, It doesn't use a for loop or start with c=p, but that's a simple way of doing a do/while that allows us to show the loop behavior in one line.)

The only choice the standard gives us to make is how to generate random bits.  The bit stream is specified as a random bit stream, but to be deterministic ours has to be pseudorandom bit stream.  A SHA256 hash chain, as you initially proposed, is not a pseudorandom byte stream because there is a deterministic algorithm for testing to see if the bit stream is the output of the generator. (block_i == SHA256(block i-1)).  I proposed one way to ensure we avoid criticism that we're not generating a pseudorandom bit stream, but there are others.